### PR TITLE
Add standalone pointsto driver

### PR DESCRIPTION
**Pull request by ziyilin**
Add standalone pointsto analysis driver class and all other necessary setting classes.

This PR is part of the standalone pointsto analysis, see https://github.com/oracle/graal/pull/4375 for details.

### CODE REVIEW

**Review by christianwimmer — CHANGES_REQUESTED** [review]
All of this new code should be in its own project `com.oracle.graal.pointsto.standalone`. That keeps the code cleanly separated from the client-independent main `pointsto` project.

**Comment by ziyilin** [comment]
> All of this new code should be in its own project `com.oracle.graal.pointsto.standalone`. That keeps the code cleanly separated from the client-independent main `pointsto` project.

@christianwimmer The code has been refactored as suggested.

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/api/PointstoOptions.java [hunk 1]
```diff
@@ -24,12 +24,12 @@
  */
 package com.oracle.graal.pointsto.api;
 
-import static jdk.vm.ci.common.JVMCIError.shouldNotReachHere;
-
 import org.graalvm.collections.EconomicMap;
 import org.graalvm.compiler.options.Option;
 import org.graalvm.compiler.options.OptionKey;
 
+import static jdk.vm.ci.common.JVMCIError.shouldNotReachHere;
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/api/PointstoOptions.java:31]: Unnecessary change

File: substratevm/mx.substratevm/suite.py [hunk 2]
```diff
@@ -1454,6 +1483,7 @@
             "dependencies": [
                 "com.oracle.svm.util",
                 "com.oracle.graal.pointsto",
+                "com.oracle.graal.pointsto.standalone",
```
Comment (christianwimmer) [inline substratevm/mx.substratevm/suite.py:1486]: We do not want the new `standalone` project in any existing distribution. At least while it is under development, we do not want to ship it in GraalVM releases.

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/HotSpotHost.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import org.graalvm.compiler.java.GraphBuilderPhase;
+import org.graalvm.compiler.nodes.graphbuilderconf.GraphBuilderConfiguration;
+import org.graalvm.compiler.nodes.graphbuilderconf.IntrinsicContext;
+import org.graalvm.compiler.options.OptionValues;
+import org.graalvm.compiler.phases.OptimisticOptimizations;
+
+import java.util.Comparator;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class HotSpotHost extends HostVM {
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/HotSpotHost.java:44]: Not a good idea to have `HotSpot` in any class name. Using "Standalone" consistently seems like the best naming convention.

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/StandaloneOptions.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package com.oracle.graal.pointsto.standalone;
+
+import org.graalvm.compiler.options.Option;
+import org.graalvm.compiler.options.OptionKey;
+
+public class StandaloneOptions {
+    @Option(help = "Specify the analysis entry point class. It should be a main Class.")//
+    public static final OptionKey<String> AnalysisEntryClass = new OptionKey<>(null);
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/StandaloneOptions.java:34]: Why is that an option, and not an argument passed to the `main` method of the standalone analysis?

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.function.Function;
+
+/**
+ * A common uninitialized field value reader class can be used by both standalone pointsto and svm
+ * hosted classes.
+ */
+public class UninitializedValueReader {
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java:42]: `UninitializedStaticFieldValueReader`

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import jdk.internal.misc.Unsafe;
+import jdk.vm.ci.meta.JavaConstant;
+import jdk.vm.ci.meta.JavaKind;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.function.Function;
+
+/**
+ * A common uninitialized field value reader class can be used by both standalone pointsto and svm
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java:39]: That's not really a useful JavaDoc comment. Since the only method is properly documented, there is no need for a class comment.

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/meta/PointsToConstantFieldProvider.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ * questions.
+ */
+
+package com.oracle.graal.pointsto.standalone.meta;
+
+import jdk.vm.ci.meta.MetaAccessProvider;
+import jdk.vm.ci.meta.ResolvedJavaField;
+import org.graalvm.compiler.core.common.spi.JavaConstantFieldProvider;
+
+public class PointsToConstantFieldProvider extends JavaConstantFieldProvider {
```
Comment (christianwimmer) [inline substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/meta/PointsToConstantFieldProvider.java:33]: Use the prefix `Standalone` instead of `PointsTo` for this class and everything else in `meta` and `plugins`.

**Review by christianwimmer — CHANGES_REQUESTED** [review]

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/api/PointstoOptions.java (hunk 1 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/api/PointstoOptions.java]: fixed

File: substratevm/mx.substratevm/suite.py (hunk 2 above)
Comment (ziyilin) [reply substratevm/mx.substratevm/suite.py]: Put it to its own jar file named standalone-pointsto.jar

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/HotSpotHost.java (hunk 3 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/HotSpotHost.java]: Fixed as suggested

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/StandaloneOptions.java (hunk 4 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/StandaloneOptions.java]: Fixed as suggested

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java (hunk 5 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java]: fixed

File: substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java (hunk 6 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/UninitializedValueReader.java]: Fixed

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/meta/PointsToConstantFieldProvider.java (hunk 7 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/meta/PointsToConstantFieldProvider.java]: Fixed. But keep the main class `com.oracle.graal.pointsto.standalone.PointsToAnalyzer` unchanged. Because it starts the pointsto analysis, and its package name already contains `standalone`.

**Review by olpaw — CHANGES_REQUESTED** [review]
There are typos in the commit message and in the PR title (statndalone)

File: substratevm/mx.substratevm/suite.py [hunk 8]
```diff
@@ -1504,6 +1533,45 @@
             },
         },
 
+        "STANDALONE_POINTSTO": {
+            "subDir": "src",
+            "description" : "A standalone version of SubstrateVM static analysis to uae for general pointsto analysis",
```
[the remark this answers is not in the document]
Comment (ziyilin) [reply substratevm/mx.substratevm/suite.py]: Thanks for pointing it out. Fixed.

**Comment by ziyilin** [comment]
> There are typos in the commit message and in the PR title (statndalone)

Fixed. Thanks.

File: compiler/mx.compiler/suite.py [hunk 9]
```diff
@@ -2047,7 +2047,7 @@
           "jdk.unsupported" # sun.misc.Unsafe
         ],
         "exports" : [
-          "* to com.oracle.graal.graal_enterprise,org.graalvm.nativeimage.pointsto,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.llvm,com.oracle.svm.svm_enterprise,com.oracle.svm_enterprise.ml_dataset,org.graalvm.nativeimage.base",
+          "* to com.oracle.graal.graal_enterprise,org.graalvm.nativeimage.pointsto,org.graalvm.nativeimage.pointsto.standalone,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.llvm,com.oracle.svm.svm_enterprise,com.oracle.svm_enterprise.ml_dataset,org.graalvm.nativeimage.base",
```
Comment (cstancu) [inline compiler/mx.compiler/suite.py:2050]: Can we avoid adding this exports to `pointsto.standalone`? It creates some problems with the module system in our CI. Could you instead use the `ModuleSuport` API like you do for other cases in `PointsToAnalyzer`?

File: substratevm/mx.substratevm/suite.py [hunk 10]
```diff
@@ -1446,8 +1475,8 @@
...
-                    "com.oracle.svm.common.option          to org.graalvm.nativeimage.pointsto,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.driver",
+                    "com.oracle.svm.util                   to org.graalvm.nativeimage.pointsto,org.graalvm.nativeimage.pointsto.standalone,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.librarysupport,org.graalvm.nativeimage.driver,org.graalvm.nativeimage.llvm,org.graalvm.nativeimage.agent.jvmtibase,org.graalvm.nativeimage.agent.tracing,org.graalvm.nativeimage.agent.diagnostics,org.graalvm.nativeimage.junitsupport,com.oracle.svm.svm_enterprise",
+                    "com.oracle.svm.common.option          to org.graalvm.nativeimage.pointsto,org.graalvm.nativeimage.pointsto.standalone,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.driver",
```
Comment (cstancu) [inline substratevm/mx.substratevm/suite.py:1479]: Can you also avoid adding this exports to `pointsto.standalone` statements?

Comment (ziyilin) [reply substratevm/mx.substratevm/suite.py]: Fixed

File: compiler/mx.compiler/suite.py (hunk 9 above)
Comment (ziyilin) [reply compiler/mx.compiler/suite.py]: Fixed by removing the export

**Review by olpaw — APPROVED** [review]

**Comment by zakkak** [comment]
Implements part of https://github.com/oracle/graal/issues/3418
